File: 5th Semester/mobile/swift/View/EditRentalView.swift
//
//  EditRentalView.swift
//  Rental
//
//  Created by Mircea Maierean on 10.11.2024.
//

import SwiftUI

struct EditRentalView: View {
    @EnvironmentObject var reservationRepository: InMemoryReservationRepository
    @EnvironmentObject var carRepository: CarRepository
    @EnvironmentObject var currentUser: User
    
    @State private var startDate: Date
    @State private var endDate: Date
    
    let reservation: Reservation
    
    init(reservation: Reservation) {
        self.reservation = reservation
        _startDate = State(initialValue: reservation.startDate)
        _endDate = State(initialValue: reservation.endDate)
    }
    
    var body: some View {
        VStack(spacing: 16) {
            if let car = carRepository.getCar(byId: reservation.carId) {
                VStack(alignment: .leading, spacing: 8) {
                    Text("Car Details")
                        .font(.title2)
                        .fontWeight(.bold)
                        .padding(.bottom, 8)
                    
                    Text("Make: \(car.make)")
                        .font(.headline)
                    
                    Text("Model: \(car.model)")
                        .font(.subheadline)
                    
                    Text("Color: \(car.color)")
                        .font(.subheadline)
                    
                    Text("Price per Day: $\(car.pricePerDay)")
                        .font(.subheadline)
                    
                    Text("Total Price: $\(calculateTotalPrice(startDate: startDate, endDate: endDate, price: car.pricePerDay))")
                        .font(.subheadline)
                        .foregroundColor(.blue)
                        .fontWeight(.semibold)
                    
                    HStack {
                        Spacer()
                    }               
                }
                .padding()
                .background(Color(UIColor.secondarySystemBackground))
                .cornerRadius(10)
                .frame(
                    minWidth: 0,
                    maxWidth: .infinity,
                    minHeight: 0,
                    maxHeight: .infinity,
                    alignment: .topLeading
                )
                .padding(.horizontal)
            } else {
                Text("Car details not found.")
                    .font(.headline)
                    .foregroundColor(.red)
            }
            
            VStack(alignment: .leading, spacing: 8) {
                Text("Start Date")
                    .font(.headline)
                
                DatePicker(
                    "Select Start Date",
                    selection: $startDate,
                    in: Date()...,
                    displayedComponents: [.date]
                )
                .onChange(of: startDate) { _ in
                    if endDate <= startDate {
                        if let nextDay = Calendar.current.date(byAdding: .day, value: 1, to: startDate) {
                            endDate = nextDay;
                        }
                    }
                }
                .padding(.horizontal)
            }
            
            VStack(alignment: .leading, spacing: 8) {
                Text("End Date")
                    .font(.headline)
                
                DatePicker(
                    "Select End Date",
                    selection: $endDate,
                    in: startDate...,
                    displayedComponents: [.date]
                )
                .padding(.horizontal)
            }
            
            Button(action: {
                if let car = carRepository.getCar(byId: reservation.carId) {
                    let updatedTotalPrice = calculateTotalPrice(
                        startDate: startDate,
                        endDate: endDate,
                        price: car.pricePerDay
                    )
                    
                    let updatedReservation = Reservation(
                        id: reservation.id,
                        carId: reservation.carId,
                        userId: reservation.userId,
                        totalPrice: updatedTotalPrice,
                        startDate: startDate,
                        endDate: endDate
                    )
                    
                    ReservationController.updateReservation(reservation: updatedReservation, repository: reservationRepository)
                }
            }) {
                Text("Save Changes")
                    .font(.headline)
                    .padding()
                    .frame(maxWidth: .infinity)
                    .background(Color.blue)
                    .foregroundColor(.white)
                    .cornerRadius(8)
            }
            .padding(.horizontal)
            .padding(.top)
            
            Spacer()
        }
        .padding()
        .navigationTitle("Edit Rental")
    }
    
    private func calculateTotalPrice(startDate: Date, endDate: Date, price: Int) -> Int {
        return price * ReservationController.totalDays(startDate: startDate, endDate: endDate)
    }
}

File: 5th Semester/mobile/localdb/rental/lib/views/edit_rental_view.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../controller/reservation_controller.dart';
import '../repository/car_repository.dart';
import '../repository/reservation_repository.dart';
import '../models/reservation.dart';

class EditRentalView extends StatefulWidget {
  final Reservation reservation;

  const EditRentalView({
    required this.reservation,
    super.key,
  });

  @override
  EditRentalViewState createState() => EditRentalViewState();
}

class EditRentalViewState extends State<EditRentalView> {
  late DateTime startDate;
  late DateTime endDate;
  bool isLoading = false;

  @override
  void initState() {
    super.initState();
    startDate = widget.reservation.startDate;
    endDate = widget.reservation.endDate;
  }

  int calculateTotalPrice({
    required DateTime startDate,
    required DateTime endDate,
    required int price,
  }) {
    final int days = endDate.difference(startDate).inDays;
    return days > 0 ? days * price : price; // Ensure at least 1 day's price
  }

  Future<void> _selectDate({
    required DateTime initialDate,
    required DateTime firstDate,
    required Function(DateTime) onDateSelected,
  }) async {
    final DateTime? picked = await showDatePicker(
      context: context,
      initialDate: initialDate,
      firstDate: firstDate,
      lastDate: DateTime(2100),
    );
    if (picked != null) {
      onDateSelected(picked);
    }
  }

  Future<void> _updateReservation(
      Reservation updatedReservation, ReservationRepository repository) async {
    setState(() {
      isLoading = true; // Show loading indicator
    });

    try {
      await ReservationController.updateReservation(
        reservation: updatedReservation,
        repository: repository,
      );

      if (mounted) {
        // Show success popup
        showDialog(
          context: context,
          builder: (context) => AlertDialog(
            title: const Text('Update Successful'),
            content: const Text('Your reservation details have been updated.'),
            actions: [
              TextButton(
                onPressed: () {
                  Navigator.of(context).pop(); // Close the dialog
                  Navigator.of(context).pop(); // Return to the previous screen
                },
                child: const Text('OK'),
              ),
            ],
          ),
        );
      }
    } catch (error) {
      if (mounted) {
        // Show failure popup
        showDialog(
          context: context,
          builder: (context) => AlertDialog(
            title: const Text('Update Failed'),
            content: Text(
              'Failed to update reservation! That\'s on our end, not yours! Please try again later.',
              style: const TextStyle(color: Colors.red),
            ),
            actions: [
              TextButton(
                onPressed: () {
                  Navigator.of(context).pop(); // Close the dialog
                },
                child: const Text('OK'),
              ),
            ],
          ),
        );
      }
    } finally {
      if (mounted) {
        setState(() {
          isLoading = false; // Hide loading indicator
        });
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    final reservationRepository =
        Provider.of<ReservationRepository>(context, listen: false);
    final carRepository = Provider.of<CarRepository>(context, listen: false);

    final car = carRepository.getCarById(widget.reservation.carId);

    return Scaffold(
      appBar: AppBar(
        title: const Text('Edit Rental'),
        centerTitle: true,
      ),
      body: Stack(
        children: [
          Padding(
            padding: const EdgeInsets.all(16.0),
            child: Center(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  Text(
                    'Your car is waiting for you',
                    style: Theme.of(context).textTheme.bodyLarge?.copyWith(
                          fontWeight: FontWeight.bold,
                        ),
                  ),
                  const SizedBox(height: 20),
                  Text(
                    'Make',
                    style: Theme.of(context)
                        .textTheme
                        .bodyMedium
                        ?.copyWith(color: Colors.grey),
                  ),
                  Text(
                    car.make,
                    style: Theme.of(context).textTheme.bodyMedium,
                  ),
                  const SizedBox(height: 8),
                  Text(
                    'Model',
                    style: Theme.of(context)
                        .textTheme
                        .bodyMedium
                        ?.copyWith(color: Colors.grey),
                  ),
                  Text(
                    car.model,
                    style: Theme.of(context).textTheme.bodyMedium,
                  ),
                  const SizedBox(height: 8),
                  Text(
                    'Year',
                    style: Theme.of(context)
                        .textTheme
                        .bodyMedium
                        ?.copyWith(color: Colors.grey),
                  ),
                  Text(
                    car.productionYear.toString(),
                    style: Theme.of(context).textTheme.bodyMedium,
                  ),
                  const SizedBox(height: 20),
                  Text(
                    'Start Date',
                    style: Theme.of(context).textTheme.bodyMedium,
                  ),
                  const SizedBox(height: 8),
                  GestureDetector(
                    onTap: () => _selectDate(
                      initialDate: startDate,
                      firstDate: DateTime.now(),
                      onDateSelected: (pickedDate) {
                        setState(() {
                          startDate = pickedDate;
                          if (endDate.isBefore(startDate)) {
                            endDate = startDate.add(const Duration(days: 1));
                          }
                        });
                      },
                    ),
                    child: Container(
                      width: double.infinity,
                      padding: const EdgeInsets.all(12),
                      decoration: BoxDecoration(
                        color: Colors.grey[200],
                        borderRadius: BorderRadius.circular(8),
                      ),
                      child: Text(
                        '${startDate.day}/${startDate.month}/${startDate.year}',
                        textAlign: TextAlign.center,
                        style: Theme.of(context).textTheme.bodyMedium,
                      ),
                    ),
                  ),
                  const SizedBox(height: 16),
                  Text(
                    'End Date',
                    style: Theme.of(context).textTheme.bodyMedium,
                  ),
                  const SizedBox(height: 8),
                  GestureDetector(
                    onTap: () => _selectDate(
                      initialDate: endDate,
                      firstDate: startDate.add(const Duration(days: 1)),
                      onDateSelected: (pickedDate) {
                        setState(() {
                          endDate = pickedDate;
                        });
                      },
                    ),
                    child: Container(
                      width: double.infinity,
                      padding: const EdgeInsets.all(12),
                      decoration: BoxDecoration(
                        color: Colors.grey[200],
                        borderRadius: BorderRadius.circular(8),
                      ),
                      child: Text(
                        '${endDate.day}/${endDate.month}/${endDate.year}',
                        textAlign: TextAlign.center,
                        style: Theme.of(context).textTheme.bodyMedium,
                      ),
                    ),
                  ),
                  const SizedBox(height: 24),
                  Text(
                    'Updated Total Price: \$${calculateTotalPrice(
                      startDate: startDate,
                      endDate: endDate,
                      price: car.pricePerDay,
                    )}',
                    style: Theme.of(context).textTheme.bodyLarge?.copyWith(
                          fontWeight: FontWeight.bold,
                          color: Colors.blue,
                        ),
                  ),
                  const SizedBox(height: 24),
                  ElevatedButton(
                    onPressed: isLoading
                        ? null
                        : () {
                            final updatedTotalPrice = calculateTotalPrice(
                              startDate: startDate,
                              endDate: endDate,
                              price: car.pricePerDay,
                            );

                            final updatedReservation = Reservation(
                              id: widget.reservation.id,
                              carId: widget.reservation.carId,
                              userId: widget.reservation.userId,
                              totalPrice: updatedTotalPrice,
                              startDate: startDate,
                              endDate: endDate,
                            );

                            _updateReservation(
                              updatedReservation,
                              reservationRepository,
                            );
                          },
                    style: ElevatedButton.styleFrom(
                      backgroundColor: Colors.black,
                      padding: const EdgeInsets.symmetric(
                          vertical: 16, horizontal: 24),
                      shape: RoundedRectangleBorder(
                        borderRadius: BorderRadius.circular(8),
                      ),
                    ),
                    child: isLoading
                        ? const CircularProgressIndicator(
                            color: Colors.white,
                          )
                        : const Text(
                            'Update Details',
                            style: TextStyle(
                              color: Colors.white,
                              fontSize: 16,
                              fontWeight: FontWeight.bold,
                            ),
                          ),
                  ),
                ],
              ),
            ),
          ),
          if (isLoading)
            const Center(
              child: CircularProgressIndicator(),
            ),
        ],
      ),
    );
  }
}
